File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int, int> frequency;

        int n = nums.size();

        for (int i = 0; i < n; i++) {
            frequency[nums[i]]++;
        }

        vector<pair<int, int>> arr;

        for (auto &p: frequency) {
            arr.push_back({p.second, p.first});
        }

        sort(arr.rbegin(), arr.rend());

        vector<int> result;
        for (int i = 0; i < k; i++) {
            result.push_back(arr[i].second);
        }

        return result;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;

        int[] result = new int[n];

        result[0] = nums[0];

        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i];
        }

        int suffixProduct = 1;

        for (int i = n - 1; i >= 1; i--) {
            result[i] = result[i - 1] * suffixProduct;
            suffixProduct *= nums[i];
        }
        result[0] = suffixProduct;

        return result;
    }
}  
